snuba/datasets/events_processor.py

```diff
@@ -0,0 +1,453 @@
+                        raise InvalidMessageType("Invalid message type: {}".format(type_))
+                    elif version == 1:
+                        if type_ in ('delete_groups', 'merge', 'unmerge'):
+                            # these didn't contain the necessary data to handle replacements
+                            return None
+                        else:
+                            raise InvalidMessageType("Invalid message type: {}".format(type_))
+                    elif version == 2:
+                        # we temporarily sent these invalid message types from Sentry
+                        if type_ in ('delete_groups', 'merge'):
+                            return None
+
+                        if type_ in ('start_delete_groups', 'start_merge', 'start_unmerge',
+                                    'start_delete_tag', 'end_delete_groups', 'end_merge',
+                                    'end_unmerge', 'end_delete_tag'):
+                            # pass raw events along to republish
+                            action_type = self.REPLACE
+                            processed = (six.text_type(event['project_id']), message)
+                        else:
+                            raise InvalidMessageType("Invalid message type: {}".format(type_))
+
+        if action_type is None:
+            raise InvalidMessageVersion("Unknown message format: " + str(message))
+
+        if processed is None:
+            return None
+
+        return (action_type, processed)
+
+    def _process_insert(self, message):
+        processed = {'deleted': 0}
+        self._extract_required(processed, message)
+
+        data = message.get('data', {})
+        # HACK: https://sentry.io/sentry/snuba/issues/802102397/
+        if not data:
+            logger.error('No data for event: %s', message, exc_info=True)
+            return None
+        self._extract_common(processed, message, data)
+
+        sdk = data.get('sdk', None) or {}
+        self._extract_sdk(processed, sdk)
+
+        tags = _as_dict_safe(data.get('tags', None))
+        self._extract_promoted_tags(processed, tags)
+
+        contexts = data.get('contexts', None) or {}
+        self._extract_promoted_contexts(processed, contexts, tags)
+
+        user = data.get('user', data.get('sentry.interfaces.User', None)) or {}
+        self._extract_user(processed, user)
+
+        geo = user.get('geo', None) or {}
+        self._extract_geo(processed, geo)
+
+        http = data.get('request', data.get('sentry.interfaces.Http', None)) or {}
+        self._extract_http(processed, http)
+
+        self._extract_extra_contexts(processed, contexts)
+        self._extract_extra_tags(processed, tags)
+
+        exception = data.get('exception', data.get('sentry.interfaces.Exception', None)) or {}
+        stacks = exception.get('values', None) or []
+        self._extract_stacktraces(processed, stacks)
+
+        return processed
+
+    def _extract_required(self, output, message):
```

Comment:
unless you want subclassses yo use this (which should not be the case) you can make them "private" (which is not really private but anyway ...) with two underscores.
Or if you need them in tests just leave them public ?

Reply:
It's not really supposed to be public, but tests are testing this directly right now. Will just make it public then.